esmvalcore/_data_finder.py

```diff
@@ -34,71 +32,42 @@ def find_files(dirnames, filenames):
-    YYYY*[-,_]*.*
-      or
-    YYYY*[-,_]YYYY*[-,_]*.*
-      or
-    YYYY*[-,_]*[-,_]YYYY*.* (Does this make sense? Is this worth catching?)
+    We assume that the filename *must* contain at least 1 year/date.
+    We look for four-to-eight-digit numbers.
+    If two potential dates are separated by either - or _,
+    we interpret this as the _daterange_ (startdate_enddate).
+    If no date range is present, but multiple potential dates are
+    present (e.g. including a version number), we assume that the last
+    number represents the date.
     """
-    name = os.path.splitext(filename)[0]
-
-    name = name.split(os.sep)[-1]
-    filename_list = [elem.split('-') for elem in name.split('_')]
-    filename_list = [elem for sublist in filename_list for elem in sublist]
-
-    pos_ydates = [elem.isdigit() and len(elem) >= 4 for elem in filename_list]
-    pos_ydates_l = list(pos_ydates)
-    pos_ydates_r = list(pos_ydates)
-
-    for ind, _ in enumerate(pos_ydates_l):
-        if ind != 0:
-            pos_ydates_l[ind] = (pos_ydates_l[ind - 1] and pos_ydates_l[ind])
-
-    for ind, _ in enumerate(pos_ydates_r):
-        if ind != 0:
-            pos_ydates_r[-ind - 1] = (pos_ydates_r[-ind]
-                                      and pos_ydates_r[-ind - 1])
-
-    dates = [
-        filename_list[ind] for ind, _ in enumerate(pos_ydates)
-        if pos_ydates_r[ind] or pos_ydates_l[ind]
-    ]
-    start_year = None
-    end_year = None
-    if len(dates) == 1:
-        start_year = int(dates[0][:4])
-        end_year = start_year
-    elif len(dates) == 2:
-        start_year, end_year = int(dates[0][:4]), int(dates[1][:4])
+    # Strip only filename from full path (and discard extension?)
+    filename = os.path.splitext(filename)[0]
+    filename = filename.split(os.sep)[-1]
+
+    # Check for a block of two potential dates separated by _ or -
+    daterange = re.findall(r'([0-9]{4,12}[-_][0-9]{4,12})', filename)
+    if daterange:
+        start_date, end_date = re.findall(r'([0-9]{4,12})', daterange[0])
+        start_year = start_date[:4]
+        end_year = end_date[:4]
     else:
-        # Slower than just parsing the name
-        try:
-            cubes = iris.load(filename)
-        except OSError:
-            raise ValueError('File {0} can not be read'.format(filename))
-
-        for cube in cubes:
-            logger.debug(cube)
-            try:
-                time = cube.coord('time')
-            except iris.exceptions.CoordinateNotFoundError:
-                continue
-            start_year = time.cell(0).point.year
-            end_year = time.cell(-1).point.year
-            break
-
-    if start_year is None or end_year is None:
-        raise ValueError(
-            'File {0} dates do not match a recognized pattern and time can '
-            'not be read from the file'.format(filename)
-        )
-    return start_year, end_year
+        dates = re.findall(r'([0-9]{4,12})', filename)
+        if not dates:
```

Comment:
If the filename does not contain dates, try reading them using iris, as the original code did. Only raise an exception if even that failed.